libnavigation-core/src/main/java/com/mapbox/navigation/core/telemetry/events/NavigationCancelEvent.kt

```diff
@@ -16,4 +17,18 @@ internal class NavigationCancelEvent(
     var comment: String = ""
 
     override fun getEventName(): String = NavigationMetrics.CANCEL_SESSION
+
+    override fun toValue(): Value {
+        val value = super.toValue()
+
+        val fields = hashMapOf<String, Value>()
+
+        arrivalTimestamp?.let { fields["arrivalTimestamp"] = it.toValue() }
+        fields["rating"] = rating.toValue()
+        fields["comment"] = comment.toValue()
+
+        (value.contents as HashMap<String, Value>).putAll(fields)
```

Comment:
It doesn't look very neat. How about you implement `toValue()` entirely in `NavigationEvent` but also add an abstract method like `fillCustomValueFields(fields: Map<String, Value>)` and in `NavigationCancelEvent` and other subclasses you only implement it? There will be no cast then.

Reply:
I'm happy to re-done it: added `protected abstract fun customFields(): Map<String, Value>?` to `NavigationEvent` class